### PR TITLE
fix(geometry): reject a zero baseline and any swapped rig in StereoCamera

### PR DESCRIPTION
**Pull request by thegoodengineer**
## What does this pull request do?

`StereoCamera`'s sign guard reads the right camera's last column, `-tx * fx`, and raised only when `torch.all(tx_fx > 0)`. Two inputs got past it that should not have:

```text
mixed batch accepted, tx: [0.5, -0.5]
tx=0 accepted, reprojected: [[[[-0.0, -0.0, -0.0]]]]
```

1. **A batch with one swapped rig was accepted.** `torch.all` fires only when every element is positive, so a batch holding one good pair and one pair with its cameras the wrong way round passed, and the swapped pair was silently reprojected behind the camera.
2. **A zero baseline was accepted.** `tx * fx == 0` is not greater than zero, so coincident cameras passed the guard. `Q` then collapses to zeros except `Q[3, 2] = -fy`, and every disparity, including a large and perfectly valid one, reprojects to exactly the origin, with no `inf` to notice.

The guard now rejects any rig whose `tx * fx` is not strictly negative. The zero case gets its own message naming `tx`, so a degenerate rig is not reported as "to be negative". The quantifier is `any`, so one bad rig cannot hide behind good ones. An empty batch stays vacuously valid, because `torch.any` of an empty tensor is `False` (`test_empty_batch_4281` still passes). The `is_exporting()` gate is unchanged.

Valid rigs, the all-negative case, take the identical path and build the identical `Q`. The only behavior change is that these two inputs now raise.

## Scope

#4270 lists four guards. This PR fixes the two above and leaves the other two alone, on purpose:

- **The `(3, 4)` shape check** was already fixed on `main` by #4385, and `test_convention_stereo_rejects_a_four_by_four_pair_4270` pins it.
- **The differing `cx`** is framed in the issue as a contract decision with two coherent outcomes: admit a differing `cx`, or drop the claim from the docs. I have not picked one. Its wart pin, `test_wart_stereo_rejects_differing_principal_points_4270`, is untouched, and so is its `#4270` reference in the class docstring and on the rst page. That is why this PR says Refs rather than Closes.

## Pins moved with the fix

Following the `grep -rnE "#4270|issues/4270|_4270"` sweep:

- `test_wart_stereo_accepts_a_batch_with_one_positive_tx_fx_4270` becomes `test_convention_stereo_rejects_a_batch_with_one_positive_tx_fx_4270`. It puts the swapped rig at both batch positions, keeps the all-positive rejection that already worked, and checks that the all-negative batch is still accepted with `tx` read back unchanged.
- `test_wart_stereo_tx_zero_collapses_every_point_to_the_origin_4270` becomes `test_convention_stereo_rejects_a_zero_baseline_4270`. It covers a single degenerate rig and one degenerate rig inside a batch, at both `0.0` and `-0.0`, since `-0.0` compares equal to zero.
- The class docstring's Convention bullet said the constructor rejects `tx` "only when it is positive (`tx = 0` and a batch with one positive product pass)". It now says strictly negative for every rig in the batch. The `.. warning::` is narrowed to the remaining `cx` part and names only its pin.
- `docs/source/geometry.camera.stereo.rst` carried the same "a zero product and a batch with one positive product pass" sentence; it now matches.

No known-failure manifest named either of the two renamed tests.

## What did you check?

The new pins fail on `main` and only where they should. I ran them from a separate worktree of `main` (with the import path verified to be that worktree):

```text
MAIN: 8 failed, 4 passed
FAILED test_convention_stereo_rejects_a_batch_with_one_positive_tx_fx_4270[cpu-float32-0] - Failed: DID NOT RAISE StereoException
FAILED test_convention_stereo_rejects_a_batch_with_one_positive_tx_fx_4270[cpu-float32-1] - Failed: DID NOT RAISE StereoException
FAILED test_convention_stereo_rejects_a_zero_baseline_4270[cpu-float32-0.0]  - Failed: DID NOT RAISE StereoException
FAILED test_convention_stereo_rejects_a_zero_baseline_4270[cpu-float32--0.0] - Failed: DID NOT RAISE StereoException
  (and the same four at float64)
```

The 4 that pass on `main` are the `cx` wart and the four-by-four convention pin, which this PR does not change.

```text
$ pytest tests/geometry/camera/test_stereo.py --dtype=float32,float64
============================== 106 passed in 0.92s ==============================

$ pytest tests/geometry/camera/test_stereo.py -k 4270 --dtype=float16,bfloat16
======================== 12 passed, 94 deselected in 0.46s ========================

$ pytest tests/geometry/camera tests/geometry/test_depth.py --dtype=float32
===================== 403 passed, 3 skipped, 22 warnings in 6.54s =====================

$ ruff check / ruff format on both changed Python files
All checks passed!
```

The export survey in `docs/export_support/cases_geomB.py` builds its stereo pair with a negative baseline (`-50.0 * 0.1`), so it is unaffected. I did not run CUDA or MPS locally.

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
## Review — #4616, head `78ab8306`

Merge onto `ecbcad561` is **clean**. Both defects reproduce exactly as you describe, the fix is the right invariant, and the pins kill the mutations that matter. One rename, then this is ready.

### [P2] The fragment should be `breaking`, not `fixed`

`changelog.d/4616.fixed.md` documents input that used to construct and now raises. One commit below the tip, #4569 ("reject a scalar magnitude past a parameter's upper bound instead of clamping it") ships that same transition as `changelog.d/4569.breaking.md`. The *content* of your fragment is already breaking-shaped — it states old behaviour, new behaviour, and the empty-batch carve-out — so this is a rename only:

```
git mv changelog.d/4616.fixed.md changelog.d/4616.breaking.md
```

(There is some precedent both ways — `4568.fixed.md` also documents a new raise — but #4569 is the closer analogue and is one commit old.)

### [P3] Stray mid-sentence line break in the rewritten docstring

`kornia/geometry/camera/stereo.py:71-72` reads:

> ... The :attr:`tx` attribute exposes the
>           negation of that
>           symbol, ...

a three-word orphan line. `ruff format` does not reflow docstring prose, so it will not self-correct. No line exceeds 120 columns, so the paragraph can simply be rewrapped.

### [P3] The new docstring clause is ungrammatical

> …: zero, a degenerate baseline, and positive, the cameras swapped, both raise.

Four nested appositives before the verb. Something like: "…strictly **negative** for every rig in the batch; a zero product (a degenerate baseline) and a positive one (the cameras swapped) both raise."

### [P3] `nan` still slips through, on both trees

`P_right[0,3] = nan` is accepted and yields `tx = nan`, because `nan == 0` and `nan > 0` are both `False`. `-inf` is also accepted (`tx = inf`); `+inf` raises. Measured on main and on the merge tree — so **not a regression**, but the inline comment ("Ensure every rig in the batch has a real baseline in the expected direction") overstates slightly. `torch.any(~(tx_fx < 0))` as a single check would reject `nan` too, but that needs its own pin and a message split, so leaving it out of scope is fine — just soften the comment.

### Verified as correct

**Both defects, reproduced on pristine `ecbcad561`:**

- Mixed batch `P_right[0,3] = [-50, +50]` → **accepted**, `tx = [0.5, -0.5]`. `reproject_disparity_to_3D(d=10)` gives `Z = 5.0` for rig 0 and `Z = -5.0` for rig 1 — the second rig reprojected behind the camera.
- `P_right[0,3] = 0` → **accepted**, `tx = [-0.]`, `Q` collapses to `[[0,0,0,-0],[0,0,0,-0],[0,0,0,0],[0,0,-100,0]]`, and `reproject(d=10)` returns `[[[[-0., -0., -0.]]]]`.

On the merge tree both raise with distinct messages, an all-negative batch is still accepted, and a valid rig's `Q` is **byte-identical to main**.

**The sign convention is the right invariant.** `tx = -P_right[0,3] / fx` and `Z = fx * tx / d`, so a physically forward point requires `tx > 0`, i.e. `P_right[0,3] < 0` — confirmed by the measured `Z = -5.0` above. This is also unchanged from main's own intent: main's message already said "Expected `T_x * f_x` to be negative"; you only changed the quantifier and added the `== 0` case. No legitimate rig's accepted sign is inverted. (The OpenCV `stereoRectify` half of that argument is *unverified by execution* — `cv2` is not installed here — but the kornia half is measured.)

**Blast radius — nothing newly red.** Whole `tests/geometry/` at `--device=cpu --dtype=float32,float64`:

| | passed | skipped | xfail | xpass | failed |
| --- | --- | --- | --- | --- | --- |
| merge tree | 5567 | 126 | 21 | 12 | **0** |
| pristine main | 5563 | 126 | 21 | 12 | **0** |

The `+4` is exactly the two renamed pins gaining a `parametrize`. `test_stereo.py --dtype=float16,bfloat16`: 18 failed on the merge tree and **18 failed on main** — identical pre-existing manifest failures, none new; the `-k 4270` subset is 12/12 green on both half dtypes. `--device=mps --dtype=float32`: 53 passed.

**No other call sites.** `StereoCamera` appears only in `kornia/geometry/camera/__init__.py`, `tests/geometry/camera/test_stereo.py`, `tests/api_surface.json` and `docs/export_support/cases_geomB.py` — and that fixture builds `right34[:, 0, 3] = -50.0 * 0.1`, negative, so it is unaffected.

**Empty batch, executed.** `torch.any(torch.zeros(0) > 0)` and `torch.any(torch.zeros(0) == 0)` are both `False`; `StereoCamera(torch.zeros(0,3,4), torch.zeros(0,3,4))` is accepted on both trees with `tx.shape == (0,)`. Dropping main's `tx_fx.numel() > 0` short-circuit is safe, and M4 below proves the carve-out is genuinely pinned rather than only asserted in the PR body.

**Convention debt is discharged for the part you fix.** On the merge tree `grep -rnE "#4270|issues/4270" kornia/` returns exactly one hit — `stereo.py:93`, the differing-`cx` question you explicitly leave open. `tests/` keeps `test_wart_stereo_rejects_differing_principal_points_4270` untouched plus your two renamed `test_convention_*` pins. Neither renamed node ID appears in `testing/known_failure_xfails/`. No commit message carries a closing keyword, so #4270 correctly stays open.

**Docs moved in lockstep.** `docs/source/geometry.camera.stereo.rst:128` no longer claims "a zero product and a batch with one positive product pass", and the `#4270` link at :130 is now only about `cx`.

**Half precision is an improvement, not a regression.** In float16, `P_right[0,3] = -1e-8` *stores* as `-0.0`, so the rig genuinely has no baseline and now raises instead of silently producing the collapsed `Q`. `-1e-5` and `-1e-3` are still accepted in fp16 and bf16; bf16 keeps `-1e-8` as a real value and accepts it. No realistic rig is newly rejected.

**Gradients and export untouched.** `StereoCamera(left, right).Q.sum().backward()` still populates `right.grad`; the `is_exporting()` gate wraps both checks, so export skips them exactly as before.

**CI: 32 SUCCESS, 9 still queued, 0 failures** — including `docs`, readthedocs, `typecheck`, `dynamo`, `tests-mps`, both half-precision legs, and three `2.5.1` legs. Pinned `ruff@0.16.7` check and `format --diff` clean.

### Mutation testing

Baseline `test_stereo.py --dtype=float32,float64` → 106 passed.

| mutation | result |
| --- | --- |
| M1 `any(tx_fx > 0)` → `all(tx_fx > 0)` (main's quantifier) | **killed** — 6 failed |
| M2 delete the `if torch.any(tx_fx == 0):` branch | **killed** — 4 failed |
| M1+M2 (= main's guard byte-for-byte) | 8 failed, 4 passed on `-k 4270` — reproduces your claim exactly |
| M3 `> 0` → `>= 0` | survived, but **equivalent**: the `== 0` branch raises first, so the two can never disagree. Not a gap. |
| M4 `any(tx_fx == 0)` → `all(tx_fx == 0)` | **killed** — 6 failed, including both `test_empty_batch_4281` nodes |

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by thegoodengineer** [comment]
All four addressed, new head `f1142cc4`. No change to the guard, the pins or the rst page.

**[P2] Fragment is now `breaking`.** `git mv changelog.d/4616.fixed.md changelog.d/4616.breaking.md`, text unchanged. Agreed on the reasoning: the change turns input that used to construct into a raise, and #4569 is the closer and more recent analogue. Thanks for pointing at it rather than just asserting the category.

**[P3] Orphan line rewrapped.** The paragraph is one wrap now, no line over 120 columns.

**[P3] Clause reworded**, essentially your version:

> which the constructor requires to be strictly **negative** for every rig in the batch; a zero product (a degenerate baseline) and a positive one (the cameras swapped) both raise.

**[P3] Comment softened, and I reproduced the non-finite behaviour first rather than take it on faith.** On this head:

```text
nan   -> accepted, tx = [nan]
-inf  -> accepted, tx = [inf]
+inf  -> StereoException Expected :math:`T_x * f_x` to be negative for every camera pair.
-50   -> accepted, tx = [0.5]
```

So the old wording, "Ensure every rig in the batch has a real baseline in the expected direction", was indeed wrong about `nan` and `-inf`. The comment now says which inputs the two comparisons do not screen and that this is unchanged from before, so the next reader is not misled into thinking non-finite input is handled:

> Neither comparison screens non-finite input: ``nan`` fails both of them and ``-inf`` is negative, so both are still accepted, exactly as they were before these checks were tightened.

I left the screening itself out of scope, as you suggested: `torch.any(~(tx_fx < 0))` would fold all three cases into one check, but it would take the zero case's dedicated message with it, and a `nan` rejection deserves its own pin.

`test_stereo.py --dtype=float32,float64`: 106 passed. Ruff check and format clean. The branch still merges cleanly, so I have not rebased it.

Thanks for the mutation table, in particular M4 catching that `any` to `all` on the zero branch takes `test_empty_batch_4281` down with it. That is the carve-out actually being pinned rather than only claimed in the body, which is what I had wanted to show and had not.

**Review by ducha-aiki — APPROVED** [review]
APPROVE on `f1142cc417fdf40c1c374776a8240f29f771a886`.

No findings. The updated head addresses every prior request: the fragment is now breaking, both prose problems are fixed, and the non-finite-input limitation is stated accurately. The zero-baseline and mixed-batch guards remain correctly pinned.

Validation: `tests/geometry/camera/test_stereo.py --device=cpu --dtype=float32,float64` passed on both the exact head and its clean synthetic merge with current `main` (`e428628d0c63e48fd983098bceb647fc799fc433`): `106 passed`. All reported CI checks pass.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5.6-sol).
